src/lifecycle.js

```diff
@@ -124,7 +129,14 @@ function triggerAttributeChanged(target, component, data) {
   var type;
   var name = data.name;
   var newValue = data.newValue;
-  var oldValue = data.oldValue;
+
+  // Read the old attribute value from cache if needed, otherwise use native oldValue
+  var cachedAttributes;
+  if (needsAttrCaching) {
+    cachedAttributes = hasOwn(target, '_cachedAttrs') && target._cachedAttrs;
```

Comment:
Instead of using a property on the element, you can use the imported `data()` function already in this scope. Something like:

```js
var cachedAtributes = data(target, 'cachedAttributes');
```

It creates an object if one doesn't already exist so you don't need to create the object if it doesn't exist in `addAttributeListeners()`.